src/main/java/io/jenkins/plugins/ml/FileParser.java

```diff
@@ -0,0 +1,143 @@
+import hudson.tasks.BuildWrapper;
+import hudson.tasks.BuildWrapperDescriptor;
+import io.jenkins.plugins.ml.model.ParsableFile;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+
+import javax.annotation.Nonnull;
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Logger;
+
+public class FileParser extends BuildWrapper {
+    public   List<ParsableFile> parsableFiles;
+    private List<String> copiedFiles = new ArrayList<>();
+
+    private static final Logger LOGGER = Logger.getLogger(FileParser.class.getName());
+
+    @DataBoundConstructor
+    public FileParser(ArrayList<ParsableFile> parsableFiles){
+        this.parsableFiles = parsableFiles;
+    }
+    @DataBoundSetter
+    public void setParsableFiles(List<ParsableFile> parsableFiles){
+        this.parsableFiles = parsableFiles;
+    }
+
+    @Override
+    public Environment setUp(AbstractBuild build, final Launcher launcher,
+                             BuildListener listener) throws IOException, InterruptedException
+    {
+        // Get the workspace
+        LOGGER.info("Recognize project workspace and folder");
+        FilePath projectWorkspace = build.getWorkspace();
+
+        // Get remote workspace if local workspace is null
+        if(projectWorkspace == null && build.getWorkspace().getRemote() != null){
+            projectWorkspace = new FilePath(new File((build.getWorkspace()).getRemote()));
+        }
+
+        // Copy each file to the workspace
+        for (ParsableFile file :  parsableFiles) {
+            FilePath copyFrom = new FilePath(new File(file.getFilePath()));
+
+
+            LOGGER.info("Copying file from " + copyFrom.toURI()
+                    + " to " + projectWorkspace.toURI());
+            String copyToFile = file.getFilePath().substring(file.getFilePath().lastIndexOf("/")+1);
+            try{
+                copyFrom.copyTo(new FilePath(projectWorkspace,copyToFile));
+                LOGGER.info("Saving name");
+                saveName(copyToFile);
+            }catch (IOException e){
+                throw new AbortException(e.getMessage());
+            }
+
+
+        }
+
+        return new Environment() {
+          @Override
+          public boolean tearDown(AbstractBuild build, BuildListener listener)
+              throws IOException, InterruptedException {
+
+            // Delete the file after build if necessary
+            for (int i = 0; i < parsableFiles.size(); i++) {
+                ParsableFile file = parsableFiles.get(i);
+                if (file != null && copiedFiles!= null && file.isDeleteFilesAfterBuild()) {
+                    FilePath projectWorkspace = build.getWorkspace();
+                    assert projectWorkspace != null;
+                    FilePath child = new FilePath(projectWorkspace, copiedFiles.get(i));
+                    child.delete();
+                }
+            }
```

Comment:
Maybe we can optimize a bit this `for` @Loghijiaha . It could be a Java8 stream expression perhaps?

---

```
                    FilePath projectWorkspace = build.getWorkspace();
                    assert projectWorkspace != null;
```

This is happening for each iteration, which doesn't seem necessary.

---

```
&& copiedFiles!= null &&
```

Doesn't seem necessary. We are creating the `copiedFiles` collection and never setting it to `null` (it could be made `final` too in fact).

---

Finally, from what I understood, we are doing the following:

1. iterate all the parseable files
2. check if the file exists and has been marked to be deleted after the build
2.1 check too that the workspace exists
3. create a `FilePath` with workspace + `copiedFiles` array value

Maybe something like this?

```java
final Optional<FilePath> projectWorkspace = Optional.of(build.getWorkspace());
    projectWorkspace.ifPresent((workspace) -> {
        parsableFiles.stream()
            .filter(file -> file != null && file.isDeleteFilesAfterBuild())
            .forEach(file -> { ... })
    }
}
```

?